### PR TITLE
 Fixes duplicate COM ports being listed

### PR DESCRIPTION
**Pull request by DaveSchmid**
## Description

- A single instance of the ObservableCollection NanoFrameworkDevices instead of synchronizing 3 instances. There is only one COM1 per PC. This is a good reason for a singleton design pattern.
- The NanoFrameworkDevices list can replace the SerialDeviceInformation list.
- Many error handling adjustments.
- Use Polly consistently.

## Motivation and Context

Finding the cause of this error was quite simple. It didn't look like but in reality my USB cable was broken. It has a loose contact. A new cable and almost everything was fine again.
Not everything was good because there is only one COM1 per PC and not two. With wobbling on the cable I even managed to generate a whole list of COM1. This should not be the case. 
Analyzing how this is possible was less easy. It is a rather complex multithreaded process with many exceptions inbetween. I followed all the exceptions and got to know an interesting library: Polly. 
I made quite a few changes until all the errors were really found and the multiple same 
COM port no longer appears.
Now it even works with 9 devices at the same time. I tested it. I don't have more Micro USB cables.
Many of the buttons in the WPF Test app do not work properly. Also, there are still many errors that have no connection with this error. I just focused on the list of COM ports.

- Fixes nanoFramework/Home#1209

## Screenshots
Extended error situation:

After the modifications. Test with 9 devices simultaneously:

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [x] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/master/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

### CODE REVIEW

File: nanoFramework.Tools.DebugLibrary.Shared/Extensions/ConditionsStateExtensions.cs
Comment (Ellerbach) [inline nanoFramework.Tools.DebugLibrary.Shared/Extensions/ConditionsStateExtensions.cs:7]: please remove dead code :-)

File: nanoFramework.Tools.DebugLibrary.Shared/NFDevice/NanoDeviceBase.cs
Comment (Ellerbach) [inline nanoFramework.Tools.DebugLibrary.Shared/NFDevice/NanoDeviceBase.cs:8]: please remove dead code :-)

File: nanoFramework.Tools.DebugLibrary.Shared/PortSerial/DeviceWatcher.cs
Comment (Ellerbach) [inline nanoFramework.Tools.DebugLibrary.Shared/PortSerial/DeviceWatcher.cs:61]: I know this start pattern with the bool was already existing before. But maybe, while you are improving this, can you create a cancellation token instead?
So you can replace the Thread.Sleep(200) with the cancellation token wait instead. And when stopping, it will not wait for nothing.

File: nanoFramework.Tools.DebugLibrary.Shared/PortSerial/PortSerial.cs
Comment (Ellerbach) [inline nanoFramework.Tools.DebugLibrary.Shared/PortSerial/PortSerial.cs:91]: I don't see a try pattern around this open, if anything goes wrong, where is it catch?

File: nanoFramework.Tools.DebugLibrary.Shared/PortSerial/PortSerial.cs
Comment (Ellerbach) [inline nanoFramework.Tools.DebugLibrary.Shared/PortSerial/PortSerial.cs:137]: Rather than the Debug, possible to use the `OnLogMessageAvailable`?

File: nanoFramework.Tools.DebugLibrary.Shared/PortSerial/PortSerial.cs
Comment (Ellerbach) [inline nanoFramework.Tools.DebugLibrary.Shared/PortSerial/PortSerial.cs:186]: for the debug messages, either use the #if DEBUG or just remove them, it's a bit cleaner

**Review by Ellerbach — COMMENTED** [review]
Thanks a lot for the improvement and fix. Few comments. I'll let @josesimoes approve all this and comment as well.

File: nanoFramework.Tools.DebugLibrary.Shared/PortSerial/DeviceWatcher.cs
Comment (DaveSchmid) [reply nanoFramework.Tools.DebugLibrary.Shared/PortSerial/DeviceWatcher.cs:61]: I think the Polly library was incorporated into the project at a later stage of development. Therefore, many repetitions are still done manually by hand. A lot could still be refactored. I have only changed what caused me an error.
In many places you just have to wait and try again. This is exactly what Polly makes very elegant.

File: nanoFramework.Tools.DebugLibrary.Shared/PortSerial/PortSerial.cs
Comment (DaveSchmid) [reply nanoFramework.Tools.DebugLibrary.Shared/PortSerial/PortSerial.cs:91]: Yes Polly does the exception handling in the call. Polly is just the right tool for these kinds of problems.

File: nanoFramework.Tools.DebugLibrary.Shared/PortSerial/PortSerial.cs
Comment (DaveSchmid) [reply nanoFramework.Tools.DebugLibrary.Shared/PortSerial/PortSerial.cs:137]: Unfortunately, the OnLogMessageAvailable does not always work! But that's another problem.

File: nanoFramework.Tools.DebugLibrary.Shared/PortSerial/PortSerial.cs
Comment (DaveSchmid) [reply nanoFramework.Tools.DebugLibrary.Shared/PortSerial/PortSerial.cs:186]: I believe Debug.WriteLine does the same as #if DEBUG.
But it's there because of the problem with OnLogMessageAvailable.  More specifically, the problem is the subsequent EventSource.WriteEvent. It stops writing when the thread is finished. And it comes to late. All is very complicated multithreading.

File: nanoFramework.Tools.DebugLibrary.Shared/Extensions/ConditionsStateExtensions.cs
Comment (DaveSchmid) [reply nanoFramework.Tools.DebugLibrary.Shared/Extensions/ConditionsStateExtensions.cs:7]: Yes, you are right. I do.

File: nanoFramework.Tools.DebugLibrary.Shared/PortSerial/DeviceWatcher.cs
Comment (DaveSchmid) [reply nanoFramework.Tools.DebugLibrary.Shared/PortSerial/DeviceWatcher.cs:61]: For example, the AddDeviceToListAsync method in PortSerialManager.cs does the same thing twice with a random delay in between. Polly could do that too. And the method would be much leaner.
But it wasn't part of that problem.

File: nanoFramework.Tools.DebugLibrary.Shared/PortSerial/PortSerial.cs
Comment (DaveSchmid) [reply nanoFramework.Tools.DebugLibrary.Shared/PortSerial/PortSerial.cs:186]: I think @josesimoes  is an expert in this. He wrote an article about it on his blog:
https://jsimoesblog.wordpress.com/2020/05/12/debugger-output-in-c-sharp/

**Comment by josesimoes** [comment]
@DaveSchmid what a test rig you have there!! 💪🏻 😄

File: nanoFramework.Tools.DebugLibrary.Shared/WireProtocol/Engine.cs
Comment (josesimoes) [inline nanoFramework.Tools.DebugLibrary.Shared/WireProtocol/Engine.cs:923]: Why was the `Abort` removed from here?

**Review by josesimoes — COMMENTED** [review]
@DaveSchmid I like your improvements here. They sure look elegant and fixing a nasty problem. Glad that someone took the time to improve this. 💯 👍🏻 

Besides my review comments, a couple more:
1. PropertyChanged.Fody and Fody have to be reverted. The versions referenced in the project are the last ones that can be used in VS. This is because of the VS extensibility engine. It's currently limited to .NET 4.7.2. 
2.  System.IO.Ports should remain at v6. This library is used in both VS2022 and VS2019. It's not safe to presume that everyone using VS2019 has .NET 7 installed, or that we have to force them to.

**Comment by josesimoes** [comment]
@DaveSchmid addressing your comment on the buggy WPF application: yes, that's assumed. This app here has been coded "as we go" to be used as a quick and dirty sandbox to test basic changes and check the code. It's very convenient for this purposed considering that the real use (in VS extension solution), requires a lot more ceremony and work to update and even making the smallest changes. 
That's why it looks sloppy. 😉  
Feel free to improve it, of course, but I would rather suggest that you put your effort and expertise on more valuable targets. There are plenty on the project! 😅

File: nanoFramework.Tools.DebugLibrary.Shared/PortSerial/DeviceWatcher.cs
Comment (DaveSchmid) [reply nanoFramework.Tools.DebugLibrary.Shared/PortSerial/DeviceWatcher.cs:61]: Yes, I noticed how super tricky the timings are. I thought at first this must be a simple bug with the two COM1s, but no, the truth is that finding this bug was very complicated! That's why I tested with many devices at the same time. But I only have ESP32. Maybe you can test it with other devices.

File: nanoFramework.Tools.DebugLibrary.Shared/WireProtocol/Engine.cs
Comment (DaveSchmid) [reply nanoFramework.Tools.DebugLibrary.Shared/WireProtocol/Engine.cs:923]: This line with Thread.Abort is the most important change of all! This caused exceptions inside of the catch of try-catch. In Engine.IncomingMessagesListener() the brake commands caused errors. Because of the Abort, the thread is no more alive at these moment.
                catch (ThreadAbortException)
                {
	        --->    break; 
                }
Also, Thread.Abort is marked as Obselete (deprecated) and should no longer be used. And it works very well without that line.

**Comment by DaveSchmid** [comment]
> @DaveSchmid I like your improvements here. They sure look elegant and fixing a nasty problem. Glad that someone took the time to improve this. 💯 👍🏻

Thanks 

> 1. PropertyChanged.Fody and Fody have to be reverted. The versions referenced in the project are the last ones that can be used in VS. This is because of the VS extensibility engine. It's currently limited to .NET 4.7.2.

Should I delete Fody?

> 2. System.IO.Ports should remain at v6. This library is used in both VS2022 and VS2019. It's not safe to presume that everyone using VS2019 has .NET 7 installed, or that we have to force them to.

I don't quite understand? There is System.IO.Ports in version 7 for .NET 4.7.2. That's good, isn't it?

File: nanoFramework.Tools.DebugLibrary.Shared/PortSerial/DeviceWatcher.cs
Comment (josesimoes) [reply nanoFramework.Tools.DebugLibrary.Shared/PortSerial/DeviceWatcher.cs:61]: Yeah... welcome to the club of the debuggers of the debugger! 🤪 

From my experience: if this works with ESP32 it will work with all the others. ESP32 is the pickiest with timmings. And all other serial connections have better performance. USB wise.

File: nanoFramework.Tools.DebugLibrary.Shared/WireProtocol/Engine.cs
Comment (josesimoes) [reply nanoFramework.Tools.DebugLibrary.Shared/WireProtocol/Engine.cs:923]: Understood. Let's see how this works in the real thing with the VS extension.
If needed we can always revisit here.

**Comment by josesimoes** [comment]
> Should I delete Fody?

Nope! You have to revert PropertyChanged.Fody and Fody NuGets to the versions that were in the project. Can NOT update them beyond those.

**Comment by josesimoes** [comment]
> I don't quite understand? There is System.IO.Ports in version 7 for .NET 4.7.2. That's good, isn't it?

OK... if that's the case, then we should be good. I thought that the version here could be somehow tied with the .NET version...

**Comment by Ellerbach** [comment]
@josesimoes @DaveSchmid, anything missing now?

**Comment by Ellerbach** [comment]
@josesimoes @DaveSchmid, anything missing now?

**Comment by josesimoes** [comment]
I'm still seeing the same versions of  PropertyChanged.Fody and Fody there... 
These need to be reverted along with the respective XSD files before this can be merged.

**Comment by DaveSchmid** [comment]
I've been doing it for a long time. Except that I forgot the references in the test tool. But why does Git (or GitHub) make my pull yours? And why the weird date?

I hate git!

**Comment by josesimoes** [comment]
@DaveSchmid yes, git can be painfull at start... 
There are no changes from upstream so you should be able to just commit yours.

**Comment by DaveSchmid** [comment]
> @DaveSchmid yes, git can be painfull at start... There are no changes from upstream so you should be able to just commit yours.

git is a disaster! Not at the start but for years.

I have everything ready and uploaded and it compiles for me with no problem.

**Comment by DaveSchmid** [comment]
i don't know why this azure pipelines coughs?

**Review by josesimoes — APPROVED** [review]
LGTM!

**Pull request merged**
